netket/utils/group/point_group.py

```diff
@@ -0,0 +1,581 @@
+
+from .semigroup import Identity, Element
+from .group import FiniteGroup
+
+from netket.utils import HashableArray, struct
+from netket.utils.float import comparable, comparable_periodic, is_approx_int
+from netket.utils.types import Array, DType, Shape
+
+############ POINT GROUP SYMMETRY CLASS ########################################
+
+
+@struct.dataclass
+class PGSymmetry(Element):
+    """
+    An abstract group element object to geometrically describe point group symmetries.
+
+    Construction: `PGSymmetry(W,w)` with an orthogonal matrix `W` and an optional
+    translation vector `w` returns an object that maps vectors :math:`\vec x` to
+    :math:`W\vec x + \vec w`.
+
+    Internally, the transformation is stored as the affine matrix [[W,w],[0,1]],
+    such that matrix inverses and multiplication correspond to inverting and
+    multiplying the transformation itself.
+    """
+
+    _affine: Array
+    """
+    A 2D array specifying the affine transformation. It has to be of the block
+    structure [[W,w],[0,1]].
+    """
+
+    def __pre_init__(self, W: Array, w: Optional[Array] = None) -> Tuple[Tuple, Dict]:
+        W = np.asarray(W)
+        if W.ndim != 2:
+            raise ValueError("W must be a 2D matrix")
+        if W.shape[0] != W.shape[1]:
+            raise ValueError("W must be a square matrix")
+        ndim = W.shape[0]
+        if not np.allclose(W.T @ W, np.eye(ndim)):
+            raise ValueError("W must be an orthogonal matrix")
+        if w is None:
+            w = np.zeros((ndim, 1))
+        else:
+            w = np.asarray(w).reshape(ndim, 1)
+        return ((np.block([[W, w], [np.zeros(ndim), 1]]),), dict())
+
+    @property
+    def affine_matrix(self) -> Array:
+        return self._affine
+
+    @property
+    def matrix(self) -> Array:
+        return self._affine[:-1, :-1]
+
+    @property
+    def translation(self) -> Array:
+        return self._affine[:-1, -1]
+
+    @property
+    def ndim(self) -> int:
+        return self._affine.shape[0] - 1
+
+    @property
+    def is_proper(self) -> bool:
+        return np.isclose(np.linalg.det(self.matrix), 1.0)
+
+    @property
+    def is_symmorphic(self) -> bool:
+        return np.allclose(self.translation, 0.0)
+
+    def __call__(self, x):
+        return np.tensordot(x, self.matrix.T, axes=1) + self.translation
+
+    def preimage(self, x):
+        return np.tensordot(x - self.translation, self.matrix, axes=1)
+
+    def k_action(self, x):
+        # wave vectors are not translated
+        return np.tensordot(x, self.matrix.T, axes=1)
+
+    def __hash__(self):
+        return hash(HashableArray(comparable(self._affine)))
+
+    def __eq__(self, other):
+        if isinstance(other, PGSymmetry):
+            return HashableArray(comparable(self._affine)) == HashableArray(
+                comparable(other._affine)
+            )
+        else:
+            return False
+
+    def change_origin(self, origin: Array) -> "PGSymmetry":
```

Comment:
wouldn it make more sense to have a `translate` by a vector operation rather than `change_origin` ?

Reply:
I feel change origin is more natural in the context of a point group.

Reply:
Yep, it's often easiest to describe nonsymmorphic point groups in terms of what operations leave what in place (see how I implemented Fd3m).

Reply:
fine by me